File: src/main/java/com/github/sschuberth/stan/Statement.java
package com.github.sschuberth.stan;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

public class Statement implements Comparable<Statement> {
    public String filename;
    public String bankId;
    public String accountId;
    public LocalDate fromDate;
    public LocalDate toDate;
    public float balanceOld;
    public float balanceNew;
    public float sumIn;
    public float sumOut;
    public List<BookingItem> bookings = new ArrayList<>();

    public Statement(String filename, String bankId, String accountId, LocalDate fromDate, LocalDate toDate, float balanceOld, float balanceNew, float sumIn, float sumOut, List<BookingItem> bookings) {
        this.filename = filename;
        this.bankId = bankId;
        this.accountId = accountId;
        this.fromDate = fromDate;
        this.toDate = toDate;
        this.balanceOld = balanceOld;
        this.balanceNew = balanceNew;
        this.sumIn = sumIn;
        this.sumOut = sumOut;
        this.bookings = bookings;
    }

    @Override
    public int compareTo(Statement st) {
        return fromDate.compareTo(st.fromDate);
    }
}

File: lib/src/main/kotlin/model/Statement.kt
@file:UseSerializers(
    dev.schuberth.stan.utils.LocaleSerializer::class,
    dev.schuberth.stan.utils.LocalDateSerializer::class
)

package dev.schuberth.stan.model

import java.time.LocalDate
import java.util.Locale

import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers

@Serializable
@Suppress("LongParameterList")
data class Statement(
    val filename: String,
    val locale: Locale,
    val bankId: String,
    val accountId: String,
    val fromDate: LocalDate,
    val toDate: LocalDate,
    val balanceOld: Float,
    val balanceNew: Float,
    val sumIn: Float,
    val sumOut: Float,
    val bookings: List<BookingItem>
) {
    companion object {
        @JvmField
        val EMPTY = Statement(
            filename = "",
            locale = Locale.ROOT,
            bankId = "",
            accountId = "",
            fromDate = LocalDate.EPOCH,
            toDate = LocalDate.EPOCH,
            balanceOld = 0f,
            balanceNew = 0f,
            sumIn = 0f,
            sumOut = 0f,
            bookings = emptyList()
        )
    }

    init {
        require(bankId.none { it.isWhitespace() }) {
            "The bank ID must not contain any whitespace characters."
        }

        require(accountId.none { it.isWhitespace() }) {
            "The account ID must not contain any whitespace characters."
        }
    }
}
